2592 futures 翻譯

FizzyElt opened the pull request:
因為內容很長所以我分段進行，減輕 reviewer 的負擔 😃 ，全數翻譯完再 merge 即可
第一階段翻譯段落
- [X] 概要
- [X] 動機
- [X] 歷史背景

第二階段翻譯段落
- [x] Guide-level explanation
- [x] Reference-level explanation

第三階段翻譯段落
- [x] Rationale, drawbacks, and alternatives
- [x] Prior art
- [x] Unresolved questions

FizzyElt committed:
copy from en RFC md

FizzyElt committed:
translate some paragraphs

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+# 概要
+[概要]: #summary
+
+該 RFC 建議為一級的 `async`/`await` 語法穩定函式庫組件。特別是，它將穩定：
```
Just my preference. Not sure if everyone agrees on this.

`propose`: 提議 is better than 建議, since 建議 in my own Chinese sense is weaker and less formal. I also notice throughout this PR somewhere uses 提議 as well.

```suggestion
此 RFC 提議替一等公民語法 `async`/`await` 穩定相關函式庫組件。特別是，它將穩定：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+該 RFC 建議為一級的 `async`/`await` 語法穩定函式庫組件。特別是，它將穩定：
+
+- `std`級別任務系統所有的 API，例如 `std::task::*`。
```
```suggestion
- `std` 中任務系統所有的 API，例如 `std::task::*`。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+- `std`級別任務系統所有的 API，例如 `std::task::*`。
+- 核心 `Future` API，例如 `core::future::Future` 和 `std::future::Future`。
+
+它*不建議*穩定任何 `async`/`await` 語法本身，這將在單獨的步驟中提出。它也不包括已在[別處](https://github.com/rust-lang/rust/issues/55766)提出的 `Pin` API 的穩定性。
```
- Would prefer bold to italic. See https://github.com/rust-tw/book-tw/pull/6
- You don't need to translate verbatim. I found it useful to swap verbs to noun sometimes.

```suggestion
它**不提議**穩定任何 `async`/`await` 語法本身，該語法自有另外的提議。它也不包括已在[別處](https://github.com/rust-lang/rust/issues/55766)提出的 `Pin` API 的穩定性。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+它*不建議*穩定任何 `async`/`await` 語法本身，這將在單獨的步驟中提出。它也不包括已在[別處](https://github.com/rust-lang/rust/issues/55766)提出的 `Pin` API 的穩定性。
+
+這是早期 [futures RFC](https://github.com/rust-lang/rfcs/pull/2418) 的修訂和精簡版本，該版本被推延到 nightly 版本獲得更多經驗之後。
```
Better?

```suggestion
這是早期 [futures RFC](https://github.com/rust-lang/rfcs/pull/2418) 的修訂和精簡版本，該 RFC 已暫緩，直到在 nightly 版本獲得更多經驗之後。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+## 為何 `Future` 要在 `std` 之中？
+
+此 RFC 的核心動機是穩定 `async`/`await` 語法的支援機制。語法本身是在（已經合併的）配套 [RFC](https://github.com/rust-lang/rfcs/pull/2394) 中提出的，並且有一篇[部落格文章](https://aturon.github.io/tech/2018/04/24/async-borrowing/)更詳細地介紹了它的重要性。
```
Some redundancies.

```suggestion
此 RFC 的核心動機是穩定 `async`/`await` 語法的支援機制。語法本身是在（已經合併的）[配套 RFC](https://github.com/rust-lang/rfcs/pull/2394) 中提出的，並且有一篇[部落格文章](https://aturon.github.io/tech/2018/04/24/async-borrowing/)更詳細介紹它的重要性。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+此 RFC 的核心動機是穩定 `async`/`await` 語法的支援機制。語法本身是在（已經合併的）配套 [RFC](https://github.com/rust-lang/rfcs/pull/2394) 中提出的，並且有一篇[部落格文章](https://aturon.github.io/tech/2018/04/24/async-borrowing/)更詳細地介紹了它的重要性。
+
+與 closures 一樣，`async` 語法涉及到生成一個匿名型別，且實現一個關鍵特徵：`Future`。因為 `async`/`await` 需要語言級別的支援，所以底層特徵也必須是標準函式庫的一部分。因此，這個 RFC 的目標是穩定這個 `Future` 特徵和它所依賴的型別。這是我們能夠穩定 `async`/`await` 本身之前所需的最後一步。
```
More Taiwanese style. And som

```suggestion
與 closures 一樣，`async` 語法涉及到生成一個匿名型別，該型別實作一個關鍵特徵：`Future`。因為 `async`/`await` 需要語言層級的支援，所以底層特徵也必須是標準函式庫的一部分。因此，這個 RFC 的目標是穩定這個 `Future` 特徵和它所依賴的型別。這是我們能夠穩定 `async`/`await` 本身之前所需的最後一步。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+## 這一步是如何融入更大的藍圖之中？
+
+
+`async`/`await` 語法是 Rust 中最渴求的特性之一，它將對整個生態系產生重大影響。以及此處描述的 API 已在 nightly 版本可用，並於 2018 年 5 月下旬開始投入使用。
```
That's what I understand it (English is not famous for its precision. 🤷🏾‍♂️)

```suggestion
`async`/`await` 語法是 Rust 中最渴求的特性之一，它將對整個生態系產生重大影響。自 2018 年 5 月下旬以來，該語法及此處描述的 API 已可於 nightly 版本使用，並被大量採用。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+# 歷史背景
+
+為穩定而提議的 API 有著悠久的歷史：
```
I feel like it's not **為穩定而提議的 API**. It might mean a bunch of RFCs propose to stabilize some APIs. Just my interpretation. I have no idea how to translate it either. May be

```suggestion
這些被提議要穩定的 API 有著悠久的歷史：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+為穩定而提議的 API 有著悠久的歷史：
+
+- `Future` 特徵從 crate 開始； [0.1 於 2016 年 8 月發布](https://aturon.github.io/tech/2016/08/11/futures/)。該版本確立了 任務/輪詢 模型的核心思想，以及此處保留的 API 的許多其他方面。 0.1 系列持續在整個 Rust 生態系統和生產系統中大量使用。
```
- Don't afraid of using fullwidth quotation marks 
- In Taiwan, we usually use 正式環境 instead of 生產環境, though I might change as time goes by...

```suggestion
- `Future` 特徵從 futures crate 開始； [0.1 於 2016 年 8 月發布](https://aturon.github.io/tech/2016/08/11/futures/)。該版本確立了「任務/輪詢」模型的核心思想，以及此處保留的 API 的許多其他方面。 0.1 系列持續在整個 Rust 生態系和正式系統中大量使用。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- `Future` 特徵從 crate 開始； [0.1 於 2016 年 8 月發布](https://aturon.github.io/tech/2016/08/11/futures/)。該版本確立了 任務/輪詢 模型的核心思想，以及此處保留的 API 的許多其他方面。 0.1 系列持續在整個 Rust 生態系統和生產系統中大量使用。
+
+- 2018 年初，隨著 `async`/`await` 的工作開始，futures 團隊建立了一個 RFC 流程並編寫了幾個 [RFC](https://github.com/rust-lang-nursery/futures-rfcs/pulls?q=is%3Apr+is%3Aclosed)，以根據長期的社區回饋對核心 API 進行修訂。這些 RFC 最終產生了 [0.2le](https://aturon.github.io/tech/2018/02/27/futures-0-2-RC/) 版本，並於 4 月[釋出](https://aturon.github.io/tech/2018/04/06/futures2/)。
```
```suggestion
- 2018 年初，隨著 `async`/`await` 的工作開始，futures 團隊建立了一個 RFC 流程並編寫了幾個 [RFC](https://github.com/rust-lang-nursery/futures-rfcs/pulls?q=is%3Apr+is%3Aclosed)，以根據長期的社群回饋對核心 API 進行修訂。這些 RFC 最終產生了 [0.2le](https://aturon.github.io/tech/2018/02/27/futures-0-2-RC/) 版本，並於 4 月[釋出](https://aturon.github.io/tech/2018/04/06/futures2/)。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- 2018 年初，隨著 `async`/`await` 的工作開始，futures 團隊建立了一個 RFC 流程並編寫了幾個 [RFC](https://github.com/rust-lang-nursery/futures-rfcs/pulls?q=is%3Apr+is%3Aclosed)，以根據長期的社區回饋對核心 API 進行修訂。這些 RFC 最終產生了 [0.2le](https://aturon.github.io/tech/2018/02/27/futures-0-2-RC/) 版本，並於 4 月[釋出](https://aturon.github.io/tech/2018/04/06/futures2/)。
+
+- 在同一時期，@withoutboats 在支援 `async` 區塊內借用的 pinning API 上的工作已經[完成](https://boats.gitlab.io/blog/post/2018-04-06-async-await-final/)。[pinning API](https://github.com/rust-lang/rfcs/pull/2349) 改變了遊戲規則，使得支援 borrowing-across-yield 而不會使 future 的核心 API 變得不安全。
```
Game-changer is slang in English. Chinese readers might not familiar with 改變了遊戲規則. Do you have an alternative translation? Just an optional enhancement tho.

This is my patch but it still irks me.

```suggestion
- 在同一時期，@withoutboats 在支援 `async` 區塊內借用的 pinning API 上的工作已經[完成](https://boats.gitlab.io/blog/post/2018-04-06-async-await-final/)。[pinning API](https://github.com/rust-lang/rfcs/pull/2349) 顛覆了過往一切，可在不使 future 的核心 API 成為 unsafe 前提下，支援 borrowing-across-yield。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- 在同一時期，@withoutboats 在支援 `async` 區塊內借用的 pinning API 上的工作已經[完成](https://boats.gitlab.io/blog/post/2018-04-06-async-await-final/)。[pinning API](https://github.com/rust-lang/rfcs/pull/2349) 改變了遊戲規則，使得支援 borrowing-across-yield 而不會使 future 的核心 API 變得不安全。
+
+- 2018 年 4 月，一對 RFC 正式提出了 `async`/`await` 語法以及 futures API 的進一步修訂（以利用 pinning API 的優勢）；後者經歷了許多修訂，包括一個新的 [RFC](https://github.com/rust-lang/rfcs/pull/2418)。最終，語法 [RFC](https://github.com/rust-lang/rfcs/pull/2394#issuecomment-387550523) 在 5 月被合併，而 API RFC 被關閉，進一步的設計疊代將在 nightly 版本進行，隨後是一個穩定的 RFC：這個！
```
- Advantage also means "benefit" or "profit", so 好處 might fit better here.
- It's not a "stable" RFC but a "stabilization RFC". The current is read like the former. On the top of my head, sorry I have no suggestion.

```suggestion
- 2018 年 4 月，一對 RFC 正式提出了 `async`/`await` 語法以及 futures API 的進一步修訂（以利用 pinning API 帶來的好處）；後者經歷了許多修訂，包括一個新的 [RFC](https://github.com/rust-lang/rfcs/pull/2418)。最終，語法 [RFC](https://github.com/rust-lang/rfcs/pull/2394#issuecomment-387550523) 在 5 月被合併，而 API RFC 被關閉，進一步的設計疊代將在 nightly 版本進行，隨後是一個穩定的 RFC：這個！
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- 2018 年 4 月，一對 RFC 正式提出了 `async`/`await` 語法以及 futures API 的進一步修訂（以利用 pinning API 的優勢）；後者經歷了許多修訂，包括一個新的 [RFC](https://github.com/rust-lang/rfcs/pull/2418)。最終，語法 [RFC](https://github.com/rust-lang/rfcs/pull/2394#issuecomment-387550523) 在 5 月被合併，而 API RFC 被關閉，進一步的設計疊代將在 nightly 版本進行，隨後是一個穩定的 RFC：這個！
+
+- 這些 API 在 5 月底[降臨到 `std`](https://github.com/rust-lang/rust/pull/51263)。
```
Choosing 降臨 is really cool but alien to me. Do have other better words?

FizzyElt replied:
降落、帶入 🤔

CYBAI replied:
how about `引進` or `導入`? It might match the context here 🤔 (or possibly simply using `加入` would be fine as well?)

FizzyElt replied:
My final decision is `加入`

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- 從那時起，隨著我們從 API 獲得的經驗，語法、`std` API 和 futures 0.3 crate 都在同步發展。這種體驗的主要驅動者是 Google 的 Fuchsia 專案，該專案在作業系統設置中大規模使用所有這些功能。
+
+- 最近的修訂是在 8 月，其中涉及一些關於如何使 `Pin` API 有更清晰的[見解](https://boats.gitlab.io/blog/post/rethinking-pin/)。這些 API 已被[提議用於穩定化](https://github.com/rust-lang/rust/issues/55766)，以及它們作為 [`self` 型別的用途](https://github.com/rust-lang/rust/issues/55786)。
```
"Propose for stabilzation" is a step to stabilize features (let them available in stable version of Rust). RFCs were proposed to stabilize those API. Therefore, a more proper translation might be

```suggestion
- 最近的修訂是在 8 月，其中涉及一些關於如何使 `Pin` API 有更清晰的[見解](https://boats.gitlab.io/blog/post/rethinking-pin/)。這些 API 已被[提議要穩定](https://github.com/rust-lang/rust/issues/55766)，以及它們作為 [`self` 型別的用途](https://github.com/rust-lang/rust/issues/55786)。
```

Still not an optimal one. Sorry.

weihanglo commented on text/2952-futures.md:
```diff
@@ -0,0 +1,599 @@
+
+- 最近的修訂是在 8 月，其中涉及一些關於如何使 `Pin` API 有更清晰的[見解](https://boats.gitlab.io/blog/post/rethinking-pin/)。這些 API 已被[提議用於穩定化](https://github.com/rust-lang/rust/issues/55766)，以及它們作為 [`self` 型別的用途](https://github.com/rust-lang/rust/issues/55786)。
+
+- 有多個兼容層可以同時使用 futures 0.1 和 0.3。這很重要，因為它允許現有生產程式碼的*增量*遷移。
```
- 兼容 -> 相容
- 生產 -> 正式 (I am not sure about this. Looks weird in this context)

```suggestion
- 有多個相容層可以同時使用 futures 0.1 和 0.3。這很重要，因為它允許既有的正式程式碼可**增量**遷移。
```

weihanglo reviewed (commented):
A lot of review comments are my personal preference. Feel free to pick up if it seems reasonable.

FizzyElt committed:
change some word

FizzyElt committed:
change line 12

FizzyElt committed:
change line 12, 36, 48, 52

FizzyElt committed:
translate guild-level explanation

weihanglo reviewed (commented):
LGTM so far

FizzyElt committed:
reference-level explanation

FizzyElt commented:
第二階段 review 分隔線
- Guide-level explanation
- Reference-level explanation

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
-Futures can be constructed through `async` blocks or `async` functions, e.g.,
+`Future` 特徵代表了一個非同步和惰性計算，它最終可能會產生一個最終值，但不以阻塞當前執行緒為前提來達成。
 
+Future 可以通過 `async` 區塊或 `async` 函式來創建，例如
```
Better to have a newline here. Some markdown parser is bad at interpreting missing newline.

```suggestion
Future 可以通過 `async` 區塊或 `async` 函式來創建，例如

```

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
 
 This `async` function, when invoked, produces a future that represents the
 completion of reading a frame from the given socket. The function signature
 is equivalent to:
```
Leftover.

```suggestion
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
 This `async` function, when invoked, produces a future that represents the
 completion of reading a frame from the given socket. The function signature
 is equivalent to:
+這個 `async` 函式，當被呼叫時，產生一個 future，代表從给定的 socket 中讀取一個幀的完成。函式簽名等同於：
```
- Sorry for being picky again. This kind of "which clause" is not quite right the grammar in written Chinese. We call it "翻譯腔". It is understandable. Just a little weird.
- I am not quite sure the translation of _frame_ in Taiwan. 幀 is correct in Simplfied Chinese context. Could @CYBAI @wusyong help take a look?

```suggestion
當呼叫這個 `async` 函式，會產生一個 future，代表完成了從给定的 socket 中讀取一個幀。函式簽名等同於：
```

CYBAI replied:
will do (possibly tomorrow) 🙇

FizzyElt replied:
I will keep going to translate the third stage.
Have any recommendations, I will be back here.

FizzyElt replied:
In Traditional Chinese, we call "frame" is "框" https://terms.naer.edu.tw/search/?q=%22frame%22&field=ti&op=AND&match=full&q=&field=ti&op=AND&order=&num=10&show=&page=&group=&heading=#result
But I think this will make reader confuse
So, I put the English followed by the word, which I think is more appropriate

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
-`Future`s or executors.
+其他非同步函數可以 **await** 這個 future; 有關完整詳細信息，請參閱[隨附的 RFC](https://github.com/rust-lang/rfcs/pull/2394)。
+
+除了 `async fn` 定義，future 還可以使用適配器來創建，就像 `Iterator` 一樣。最初，這些適配器將完全由「out of tree」提供，但最终它們將帶入標準函式庫。
```
- Adapter is 轉接器/轉接頭 in Taiwan. I feel like more suitable translation is "轉接器".
- It's not "provided by out-of-tree"; it means "provided through out-of-tree".
- And when we are not certain about a term is widely used or not, we can always add a pair of parentheses to include the original term.

```suggestion
除了 `async fn` 定義，future 還可以使用轉接器（adapter）來創建，就像 `Iterator` 一樣。最初，這些轉接器將完全以「out of tree」形式提供，但最终會以它們自己的方式併入標準函式庫。
```

FizzyElt replied:
I'm not sure what is "tree" to mean in here
I think this means the "other crates"

weihanglo replied:
An in-tree or out-of-tree module generally refers to whether a module is in the main codebase or not. 

In rust, we usually see modules or functions residing rust-lang/rust without using git subtree/submodules as in-tree, otherwise is out-of-tree.

weihanglo replied:
I current have no better term to translate out-of-tree neither.

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
+
+最終非同步計算以**任務**的形式執行，相當於輕量級執行緒。 **executor** 提供了從 `()` 生成的 `Future` 創建任務的能力。執行器將固定 `Future` 並對其進行 `poll`，直到為其創建的任務完成。
+
+executor 的實作以合作的方式調度它擁有的任務。是否使用一個或多個作業系統執行緒以及可以在其之上並行生成多少任務取決於 executor 的實作。一些 executor 的實作可能只能驅動單個 `Future` 完成，而另一些則可以提供動態接受新 `Future` 的能力，這些 Future 是在任務中被驅動完成。
```
- For "cooperative", we generally use "協同式" in terms of translation.
- Paralle should be "平行". [ref](https://hackmd.io/@sysprog/concurrency/https%3A%2F%2Fhackmd.io%2F%40sysprog%2FS1AMIFt0D)

```suggestion
executor 的實作以協同式地（cooperative）排程它的任務。是否使用一個或多個作業系統執行緒以及可以在其之上平行生成多少任務取決於 executor 的實作。一些 executor 的實作可能只能驅動單個 `Future` 完成，而另一些則可以提供動態接受新 `Future` 的能力，這些 Future 是在任務中被驅動完成。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -59,52 +59,35 @@
+
+executor 的實作以合作的方式調度它擁有的任務。是否使用一個或多個作業系統執行緒以及可以在其之上並行生成多少任務取決於 executor 的實作。一些 executor 的實作可能只能驅動單個 `Future` 完成，而另一些則可以提供動態接受新 `Future` 的能力，這些 Future 是在任務中被驅動完成。
+
+該 RFC 不包含任何 executor 的定義。它僅以 API 的形式定義了 executors、tasks 和 `Future` 之間的整合，允許任務請求再次被調度。 `task` 模組提供了這些 API，在手動實作 `Future` 或 executor 時需要這些 API。
```
```suggestion
該 RFC 不包含任何 executor 的定義。它僅以 API 的形式定義了 executors、tasks 和 `Future` 之間的交互關係，這些 API 允許任務請求再次進行排程。 這些 API 由 `task` 模組提供，在手動實作 `Future` 或 executor 時需要這些 API。
```

weihanglo reviewed (commented):
Thanks for your continuous works on this! Pardon me being picky 🙇🏾

FizzyElt committed:
[fix] change some line
add new line in 67
remove en words 72 ~ 74
change line 84
change line 88
change line 90

FizzyElt committed:
translate third stage

FizzyElt commented:
第三階段分隔，麻煩各位了
這次可能會有較多語句不太通順的地方，我盡量以我的認知修到可順暢閱讀的程度

- Rationale, drawbacks, and alternatives
- Prior art
- Unresolved questions

weihanglo commented:
Sorry life has been way busier tst this week. I will try to find a slot to review on the weekend

weihanglo commented on text/2952-futures.md:
```diff
@@ -79,7 +79,7 @@ fn read_frame<'sock>(socket: &'sock TcpStream)
 其他非同步函數可以 **await** 這個 future; 有關完整詳細信息，請參閱[隨附的 RFC](https://github.com/rust-lang/rfcs/pull/2394)。
 
-除了 `async fn` 定義，future 還可以使用轉接器（adapter）來創建，就像 `Iterator` 一樣。最初，這些轉接器將完全由樹之外所提供，但最终它們將帶入標準函式庫。
+除了 `async fn` 定義，future 還可以使用轉接器（adapter）來創建，就像 `Iterator` 一樣。最初，這些轉接器將完全由其他 crate 所提供，但最终它們將帶入標準函式庫。
```
Brilliant! Somewhat not a 1 to 1 translation but I love it!

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
 為了彌補 futures 0.1 和 0.3 之間的差距，有幾個相容性鋪墊，包括一個內置於 futures crate 本身的，您可以簡單地透過使用 `.compat()` 組合器在兩者之間切換。這些相容層使現有生態系可以順利地使用新的 futures API，並使大型程式的漸進式轉換成為可能。
 
-# Rationale, drawbacks, and alternatives
+# 基礎原理、缺點和替代方案
```
Maybe this is sufficient?

```suggestion
# 原理、缺點和替代方案
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
-async/await syntax (and
-the
-[borrowing it supports](https://aturon.github.io/tech/2018/04/24/async-borrowing/)).
+到目前為止，我們已經能夠將「任務/輪詢」模型推向幾乎所有 Rust 希望佔據的利基點，而主要的缺點是缺乏 async/await 語法（以及它[支持的借用](https://aturon.github.io/tech/2018/04/24/async-borrowing/)）。
```
支援 is more a Taiwanese term though.

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
-atomics can be implemented as follows:
+總而言之，futures 的主要替代模型是 callback-base 的方法，在發現當前方法之前嘗試了 callback-base 幾個月。根據我們的經驗，callback 方法在 Rust 中有幾個缺點：
+
+- 它幾乎在所有地方都被強制分配，因此與 no_std 不相容。
```
This is not clear to me. Could we consider adding a supplement?
```suggestion
- 它幾乎在所有地方都強制分配（allocation），因此與 no_std 不相容。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+總而言之，futures 的主要替代模型是 callback-base 的方法，在發現當前方法之前嘗試了 callback-base 幾個月。根據我們的經驗，callback 方法在 Rust 中有幾個缺點：
+
+- 它幾乎在所有地方都被強制分配，因此與 no_std 不相容。
+- 它使取消變得**非常**困難，而對於提議的模型，它只是「刪除」。
```
I believe "drop" here refers to Drop trait. I would keep it as-was, such like:

```suggestion
- 它使取消變得**非常**困難，而對於提議的模型，刪除只是「drop」。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- 它幾乎在所有地方都被強制分配，因此與 no_std 不相容。
+- 它使取消變得**非常**困難，而對於提議的模型，它只是「刪除」。
+- 主觀上，組合器程式碼非常繁瑣，而基於任務的模型則可以輕鬆且快速的達成。
```
Either using `地` or just dropping it.

```suggestion
- 主觀上，組合器程式碼非常繁瑣，而基於任務的模型則可以輕鬆且快速達成。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+## 移除內嵌錯誤的基本原理、缺點和替代方法
+
+在主要特徵中刪除內嵌（build-in）錯誤型別有多種原因：
```
Consistent with line 356.

```suggestion
在主要特徵中移除內嵌（built-in）錯誤型別有多種原因：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+在主要特徵中刪除內嵌（build-in）錯誤型別有多種原因：
+
+- **改進的類型檢查和推斷**。錯誤型別是當今使用 futures 組合器時最大的痛點之一，無論是試圖讓不同的型別配對，還是在一段代碼無法產生錯誤時導致推理失敗。需要清楚的是，當 `async` 語法可用時，這些問題將變得不那麼明顯。
```
- 類型 -> 型別
- 代碼 -> 程式碼

```suggestion
- **改進的型別檢查和推斷**。錯誤型別是當今使用 futures 組合器時最大的痛點之一，無論是試圖讓不同的型別配對，還是在一段程式碼無法產生錯誤時導致推理失敗。更明確地說，當 `async` 語法可用時，這些問題將不再那麼明顯。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **改進的類型檢查和推斷**。錯誤型別是當今使用 futures 組合器時最大的痛點之一，無論是試圖讓不同的型別配對，還是在一段代碼無法產生錯誤時導致推理失敗。需要清楚的是，當 `async` 語法可用時，這些問題將變得不那麼明顯。
+
+- **非同步函式**。如果我們保留一個內嵌的錯誤型別，那麼 `async fn` 應該如何工作就不太清楚了：它是否應該總是要求返回的型別是一個 `Result`？如果不是，當返回非 `Result` 型別時會發生什麼？
```
```suggestion
- **非同步函式**。如果我們保留一個內嵌的錯誤型別，那麼 `async fn` 應該如何工作就不太清楚了：它是否應該總是要求回傳型別是一個 `Result`？如果不是，當回傳非 `Result` 型別時會發生什麼？
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **非同步函式**。如果我們保留一個內嵌的錯誤型別，那麼 `async fn` 應該如何工作就不太清楚了：它是否應該總是要求返回的型別是一個 `Result`？如果不是，當返回非 `Result` 型別時會發生什麼？
+
+- **組合器清晰度**。透過它們是否依賴錯誤來拆分組合器可以闡明語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
```
Not sure if it is better in this, no?

```suggestion
- **組合器清晰度**。按照組合器依賴錯誤與否來拆分之，可以闡明其語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
```

FizzyElt replied:
>Combinator clarity. Splitting up the combinators by whether they rely on errors or not clarifies the semantics. This is especially true for streams, where error handling is a common source of confusion.

I am not sure what is "they" means here.

weihanglo replied:
There is nothing here so it supposed to be combinators.

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **組合器清晰度**。透過它們是否依賴錯誤來拆分組合器可以闡明語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
+
+- **正交性**。一般來說，產生和處理錯誤與核心輪詢機制是分開的，所以在同等條件的情況下，遵循 Rust 的一般設計原則，透過與 `Result` **組合**來處理錯誤似乎很好。
```
Seems better? 

Though I don't really know how to translate orthogonality; it means two things are tangent (what am I talking about 😆) and not related to each other in this context. 正交性 is technicially correct in terms of translation while it still looks alient for people not bearing strong mathematical background.

```suggestion
- **正交性**。一般來說，產生和處理錯誤與核心輪詢機制是分開的，所以在同等條件的情況下，遵循 Rust 的一般設計原則，藉由**組合** `Result` 來處理錯誤似乎更好。
```

FizzyElt replied:
I add the reference in here to help reader understand.
Wiki "Orthogonality in programming"
https://en.wikipedia.org/wiki/Orthogonality_(programming)

weihanglo commented on text/2952-futures.md:
```diff
@@ -485,36 +414,18 @@ static VTABLE: &RawWakerVTable = &RawWakerVTable {
+
+以 async/await 語法和以 future（又名 promise）為基礎的大量現有技術。提議的 futures API 是受到 Scala 的 futures 的影響，並且與各種其他語言的 API 大體相似（就以提供的轉接器而言）。
+
+此 RFC 中模型的獨特之處在於使用了任務，而不是 callback。RFC 的作者不知道其他 *future* 函式庫是否使用這種技術，但它是 functional programming 中一種相當知名且更普遍的技術。有關於最近的範例，請參閱這篇關於 Haskell 並行的[論文](https://www.microsoft.com/en-us/research/wp-content/uploads/2011/01/monad-par.pdf)。這個 RFC 似乎是一个將 「trampoline」 技術語明確的、開放式的任務/換醒模型结合起来的新想法。
```
I believe "parallelism" is 平行 but I am a bit lazy to find a reference. Calling @CYBAI 😈 

```suggestion
此 RFC 中模型的獨特之處在於使用了任務，而不是 callback。RFC 的作者不知道其他 *future* 函式庫是否使用這種技術，但它是 functional programming 中一種相當知名且更普遍的技術。有關於最近的範例，請參閱這篇關於 Haskell 平行的[論文](https://www.microsoft.com/en-us/research/wp-content/uploads/2011/01/monad-par.pdf)。這個 RFC 似乎是一个將 「trampoline」 技術與明確的、開放式的任務/喚醒模型结合起来的新想法。
```

FizzyElt replied:
I found it 😄 
https://terms.naer.edu.tw/search/?q=parallelism&field=ti&op=AND&group=1&num=10&order=cat&page=3

CYBAI replied:
Yes, "parallelism" is 平行 and "concurrency" is 並行. We can refer to jserv's materials.

> 光看翻譯，Concurrency (並行) 和 Parallelism (平行) 已令許多人困惑，多執行緒程式設計更是工程人員的夢魘。本講座希望帶著學員回歸本質，並透過案例分析來掌握並行程式設計。

ref: https://hackmd.io/@sysprog/concurrency/https%3A%2F%2Fhackmd.io%2F%40sysprog%2FS1AMIFt0D

weihanglo commented on text/2952-futures.md:
```diff
@@ -485,36 +414,18 @@ static VTABLE: &RawWakerVTable = &RawWakerVTable {
-None at the moment.
+雖然此解決方案向 `LocalWaker` 方法提供了較差的錯誤消息（因為在錯誤的執行緒上發生 `wake` 之前，它不会恐慌，而不是在 `LocalWaker` 轉換為 `Waker` 時發生恐慌），它極大地透過去除 `LocalWaker` 和 `Waker` 型別的重複，簡化了面向用戶的 API 。
+
+在實際中，Rust 生態系統中最常見的執行程序也很可能繼續與多執行緒相容（就像現在一樣），因此在更專業的情況下，針對這種情況的人因工程學進行優化優先於更好的錯誤訊息。
```
- I never heard "在實際中". Perhaps "實際上"？
- Executor is a proper noun. I'd prefer to translate it into 執行器 or leave it as-is.

```suggestion
在實際情況下，非常可能 Rust 生態系統中最常見的執行器繼續與多執行緒相容（就像現在一樣），因此在更特化的情況，針對這種情況最佳化其人因工程學，優先於更好的錯誤訊息。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -485,36 +414,18 @@ static VTABLE: &RawWakerVTable = &RawWakerVTable {
-[unresolved]: #unresolved-questions
-
-None at the moment.
+雖然此解決方案向 `LocalWaker` 方法提供了較差的錯誤消息（因為在錯誤的執行緒上發生 `wake` 之前，它不会恐慌，而不是在 `LocalWaker` 轉換為 `Waker` 時發生恐慌），它極大地透過去除 `LocalWaker` 和 `Waker` 型別的重複，簡化了面向用戶的 API 。
```
Technically user was 使用者 in Taiwanese translation convention in old days.

```suggestion
雖然此解決方案向 `LocalWaker` 方法提供了較差的錯誤訊息（因為在錯誤的執行緒上發生 `wake` 之前，它不会恐慌，而不是在 `LocalWaker` 轉換為 `Waker` 時發生恐慌），它極大地透過去除 `LocalWaker` 和 `Waker` 型別的重複，簡化了給使用者的 API 。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **正交性**。一般來說，產生和處理錯誤與核心輪詢機制是分開的，所以在同等條件的情況下，遵循 Rust 的一般設計原則，透過與 `Result` **組合**來處理錯誤似乎很好。
+
+綜上所述，即使使用 `TryFuture`，錯誤嚴重的程式碼也有真正的缺點：
```
Error-heavy means using Error a lot IMO, not severely error'd.

```suggestion
綜上所述，即使使用 `TryFuture`，對大量涉及錯誤處理的程式碼來說，仍有真正的缺點：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+綜上所述，即使使用 `TryFuture`，錯誤嚴重的程式碼也有真正的缺點：
+
+- 需要額外的引入（如果程式碼引入 future prelude，則可以避免，我們也許可以更直接地鼓勵你這麼做）。
```
```suggestion
- 需要額外的引入（如果程式碼引入 future prelude，則可以避免，我們也許可以更直接地鼓勵這麼做）。
```

weihanglo reviewed (commented):
Again, thank you for putting in this! 

The followings two paragraphs have yet completely reviewed.

- [x] Rationale, drawbacks and alternatives for removing built-in errors
- [x] Rationale, drawbacks and alternatives to the core trait design (wrt `Pin`)

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- 需要額外的引入（如果程式碼引入 future prelude，則可以避免，我們也許可以更直接地鼓勵你這麼做）。
+
+- 對於程式來說，受一個特徵的**約束**而**實作**另一個特徵可能會令人困惑。
```
```suggestion
- 從程式碼角度來說，受一個特徵的**約束**卻**實作**另一個特徵可能會令人困惑。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+## 核心特徵設計的基本原理、缺點和替代方案 (wrt `Pin`)
+
+撇開上面討論過的正交錯誤處理不談，這個 RFC 中主要的另一個重要項目是核心輪詢方法的 `Pin` 轉移，以及它與 `Unpin`/手動撰寫的 futures 之間的關係。在 RFC 討論過程中，我們基本上確定了解決這個問題的三種主要方法：
```
```suggestion
撇開上面討論過的錯誤處理之正交性不談，這個 RFC 中的另一主要項目是核心輪詢方法向 `Pin` 轉移，以及它與 `Unpin`/手動撰寫的 futures 之間的關係。在 RFC 討論過程中，我們基本上確定了解決這個問題的三種主要方法：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **兩個核心特徵**。我們可以提供兩個特徵，例如 `MoveFuture` 和 `Future`，其中一個在 `&mut self` 上運行，另一個在 `Pin<&mut Self>` 上運行。這使得我們可以繼續以 futures 0.2 的風格編寫程式，即無需引入 `Pin`/`Unpin` 或以其他方式談論 pin。一個關鍵要求是需要互操作性（interoperation），以便可以在任何需要 `Future` 的地方使用 `MoveFuture`。至少有兩種方法可以實現這種互操作：
+
+  - 透過對 `T: MoveFuture` 的 `Future` 全面實現（blanket implementation）。這種方法目前阻止了一些**其他**所需的實現（特別圍繞在 `Box` 和 `&mut`），但這似乎不是問題的根本。
```
The curren term we chose was 毯子實作, which seems not correct at this moment. We can also fix that in TRPL translation.

https://rust-lang.tw/book-tw/ch10-02-traits.html?highlight=blanket#透過特徵界限來選擇性實作方法

> #### blanket
> 
> adjective _[attributive]_
> **covering all cases or instances; total and inclusive:** _a blanket ban on tobacco advertising._


```suggestion
- 透過對 `T: MoveFuture` 的 `Future` 全面實作（blanket implementation）。這種方法目前阻止了一些**其他**所需的實作（特別圍繞在 `Box` 和 `&mut`），但這似乎不是問題的根本。
```

cc @wusyong

FizzyElt replied:
I think we can give a reference, the reader can figure out what this means.
Like this example
https://github.com/pretzelhammer/rust-blog/blob/master/posts/tour-of-rusts-standard-library-traits.md#generic-blanket-impls

weihanglo replied:
We could but not necessary at this moment. If we substitute **全面實作** for **毯子實作** in TRPL zh-TW version, people can search and figure out themselves.

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+另一方面，雙特徵方法也有缺點。如果我們**還**刪除錯誤型別，則會出現組合爆炸，因為我們最終需要每個特徵的 `Try` 變體（這也延伸到相關的特徵，例如 `Stream`）。更廣泛地說，使用單一特徵方法，`Unpin` 充當一種「獨立旋鈕」，可以與其他關注點作正交應用；使用雙特徵方法，它是「混合的」。目前這兩種雙特徵方法都受到了編譯器的限制，儘管這不應該被視為決定因素。
+
+**該 RFC 選擇單一特徵方法的主要原因是它是保守的、前向相容的選擇，並且已經在實際中證明了自己**。可以在未來的任何時候添加 `MoveFuture` 以及一個全面實現。因此，從本 RFC 中提出的單一 `Future` 特徵開始，在我們獲得經驗的同時，我們的選項也保持最大限度的開放。
```
I always confused by translation of different kinds of compatibilities. Could we attach the original term here?

```suggestion
**該 RFC 選擇單一特徵方法的主要原因是它是保守的、前向相容（forward-compatible）的選擇，並且已經在實際中證明了自己**。可以在未來的任何時候添加 `MoveFuture` 以及一個全面實作。因此，從本 RFC 中提出的單一 `Future` 特徵開始，在我們獲得經驗的同時，我們的選項也保持最大限度的開放。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+該 RFC 採用的「一個核心特徵」方法的缺點是，它在手動編寫可移動 future 時與人因工程相衝突：您現在需要為您的型別引入 `Pin` 和 `Unpin`、調用 `poll_unpin` 和實現 `Unpin`。這一切都很機械化，但這是痛苦的。而 `Pin` 的人因工程學的改進可能會消除其中一些問題，但仍存在許多未解決的問題。
+
+另一方面，雙特徵方法也有缺點。如果我們**還**刪除錯誤型別，則會出現組合爆炸，因為我們最終需要每個特徵的 `Try` 變體（這也延伸到相關的特徵，例如 `Stream`）。更廣泛地說，使用單一特徵方法，`Unpin` 充當一種「獨立旋鈕」，可以與其他關注點作正交應用；使用雙特徵方法，它是「混合的」。目前這兩種雙特徵方法都受到了編譯器的限制，儘管這不應該被視為決定因素。
```
```suggestion
另一方面，雙特徵方法也有缺點。如果我們**同時**移除錯誤型別，則會出現組合爆炸，因為我們最終需要每個特徵的 `Try` 變體（這也延伸到相關的特徵，例如 `Stream`）。更廣泛地說，使用單一特徵方法，`Unpin` 充當一種「獨立旋鈕」，可以與其他關注點作正交應用；使用雙特徵方法，它是「混合的」。目前這兩種雙特徵方法都受到了編譯器的限制，儘管這不應該被視為決定因素。
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+## 喚醒設計 (Waker) 的基本原理、缺點和替代方案
+
+該提議的先前疊代包括一個單獨的喚醒型別 `LocalWaker`，它是 `!Send + !Sync` 並且可用於實現優化的執行器行為，而無需原子引用計數或原子喚醒。然而，實際上，這些相同的優化可以透過使用執行緒本地喚醒隊列來實現，攜帶 ID 而不是指向喚醒物件的指標，並跟蹤執行程序 ID 或線程 ID 以執行未發送 `Waker` 的 runtime 斷言執行緒。舉個簡單的例子，零原子的單執行緒鎖定執行器可以如下實現：
```
```suggestion
該提議的先前疊代包括一個單獨的喚醒型別 `LocalWaker`，它是 `!Send + !Sync` 並且可用於實作最佳化的執行器行為，而無需原子引用計數或原子喚醒。然而，實際上，這些相同的最佳化可以使用執行緒本地喚醒佇列（queue）來實作，攜帶 ID 而不是指向喚醒物件的指標，並追蹤執行程序 ID 或執行緒 ID 以執行未發送 `Waker` 的 runtime 斷言執行緒。舉個簡單的例子，零原子的單執行緒鎖定執行器可以如下實現：
```

weihanglo commented on text/2952-futures.md:
```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
+
+- **一個核心特徵**。這就是主要 RFC 文本中採用的方法：只有一個核心 `Future` 特徵，它適用於 `Pin<&mut Self>`。另外還有一個 `poll_unpin` 輔助器，用於在手動實現中使用 `Unpin` futures。
+
+- **兩個核心特徵**。我們可以提供兩個特徵，例如 `MoveFuture` 和 `Future`，其中一個在 `&mut self` 上運行，另一個在 `Pin<&mut Self>` 上運行。這使得我們可以繼續以 futures 0.2 的風格編寫程式，即無需引入 `Pin`/`Unpin` 或以其他方式談論 pin。一個關鍵要求是需要互操作性（interoperation），以便可以在任何需要 `Future` 的地方使用 `MoveFuture`。至少有兩種方法可以實現這種互操作：
```
Not sure about **互操作性**. 協同工作性 seems not good either. Do anyone get a reference for it?

FizzyElt replied:
"相互運作" is colloquialism.
https://terms.naer.edu.tw/search/?q=interoperation&field=ti&op=AND&match=&q=&field=ti&op=AND&order=cat&num=10&show=&page=5&group=1&heading=#result

dieterplex replied:
It seems 互操作性 is widely accepted in simplified chinese. I think it's reasonable since we usually using  "互動的" <---> "interactive". I would prefer 可交互運作性 here.
https://terms.naer.edu.tw/detail/3263013/ 可互操作性 interoperation
https://terms.naer.edu.tw/detail/2914181/ 互通性 Interoperability

weihanglo reviewed (commented):
Not a blocker. Just my personal taste favouring the old way for Taiwanese terms.

- [ ] Replace 優化 with 最佳化 throughout.
- [ ] Replace 實現 with 實作 throughout.

FizzyElt committed:
three stage change

FizzyElt committed:
Update the omitted issue

FizzyElt commented:
I discover the file name is not correct.
So, if ready for merge, please notice me to change file name.

FizzyElt committed:
change some words and sentences

weihanglo reviewed (approved):
Thank you for your great effort so far! This translation is really incredible. Definitely help a lot of people in our community.

FizzyElt committed:
change the file numbering

FizzyElt commented:
I changed the file numbering.
You can merge directly if everything is ready.
